nav2_collision_monitor: collision detector

tonynajjar opened the pull request:
<!-- Please fill out the following pull request template for non-trivial changes to help us process your PR faster and more efficiently.-->

---

## Basic Info

| Info | Please fill out this column |
| ------ | ----------- |
| Ticket(s) this addresses   | https://github.com/ros-planning/navigation2/issues/3184 |
| Primary OS tested on | Ubuntu |
| Robotic platform tested on | (Steve's Robot, gazebo simulation of Tally, hardware turtlebot) |

---

## Description of contribution in a few bullet points

This PR introduces the Collision detector, a node similar to the Collision Monitor in the sense that it detects obstacles in regions, with the difference that the CD does not affect the velocity directly, but instead publishes the detections to a topic for other Nodes in the system to take actions

## Description of documentation updates required from your changes

<!--
* Added new parameter, so need to add that to default configs and documentation page
* I added some capabilities, need to document them
-->

---


#### For Maintainers: <!-- DO NOT EDIT OR REMOVE -->
- [ ] Check that any new parameters added are updated in navigation.ros.org
- [ ] Check that any significant change is added to the migration guide
- [ ] Check that any new features **OR** changes to existing behaviors are reflected in the tuning guide
- [ ] Check that any new functions have Doxygen added
- [ ] Check that any new features have test coverage
- [ ] Check that any new plugins is added to the plugins page
- [ ] If BT Node, Additionally: add to BT's XML index of nodes for groot, BT package's readme table, and BT library lists

tonynajjar commented:
Happy to hear your opinions @SteveMacenski @AlexeyMerzlyakov

SteveMacenski commented:
From a quick glance, I think the discussed solution(s) in https://github.com/ros-planning/navigation2/issues/3496  are better fits. This seems less contextual with just the boolean and also a bit odd to have it be an "action" like slow/stop are actions. I think it would be sensible to have this streaming state at its natural frequency (even with `None`s) which makes the added state for `"publish"` not really relevant. 

Or if not published at the natural rate, we already have a check for `printAction` if the state changes, we can change that to `notifyActionChanges` and log + publish. No need to embed this deeper into the polygons or the action processing logic :smile:

tonynajjar commented:
https://github.com/ros-planning/navigation2/issues/3496 would not fully satisfy my requirements. I need a polygon with an action that will not affect cmd_vel, maybe the DO_NOTHING action could be used for it? Then by publishing the "state" of collision monitor as explained in the #3496, I would get the information whether my DO_NOTHING polygon is triggered or not.

One more requirement I have is that this DO_NOTHING polygon should run based on a timer and not in the subscription of cmd_vel (We want to check for a "collision" while the robot is stationary). I workaround it currently by doing exactly that, creating a timer a processing the DO_NOTHING/PUBLISH polygons in it. (not shown in this PR). I don't love this workaround though

AlexeyMerzlyakov commented:
@tonynajjar, I think the best solution here should be to add new `<polygon_name>.do_nothing` ROS2-option to the polygon properties. This change plus #3496 should meet the Req.1 (Just report state when polygon was triggered). This is much more unified solution, since it will also cover the situations when we need to track "approaching to the obstacle" state changes, or the case when we have to set a specific polygon with minimal number of points triggering the specific action.

Regarding Req.2 (publishing the state on regular rate), it could be done either by CM or by operating s/w side. However, from the performance perspective, CM is better to publish the state one time since it was changed (the same as all other Nav2 nodes currently do). Adding the rate to the CM, I do not consider as a good choice since: 1. as mentioned above, performance reasosns, 2. in some cases, operating s/w should know as soon as possible the state change. You can add the message callback, and set internal state in your application to "on"/"off"; and then frequently query this status, e.g. on regular `rclcpp::Rate` basis in this case.

SteveMacenski commented:
@tonynajjar this seems like you need a completely separate addition to this server, if it makes sense to add here. Can you provide some specific information about what you're trying to do so it can help drive our mental model of how to help? 

If you just need <something> published on a timer, we could just create a timer to publish <stuff> separately of any of this within the server. I don't understand what it is you're looking for (or why) to give a more concrete suggestion. But I don't think adding another polygon type for "do nothing" is a great answer (but might be right?)

tonynajjar commented:
Alright let me make it clearer.

Simply put, I need a node that will publish (e.g. a bool) whether a data point (e.g. from a pointcloud) is inside a polygon. So it has nothing to do with cmd_vel, and therefore has to run on a timer, i.e. even when the robot is stopped.

In case you're wondering, the exact application for us is to detect objects below the forks of our robot and notify about them (no cmd_vel involved).

If this should even be in Nav2 is a valid question, but I felt like nav2_collision_monitor is too close to accomplishing that it didn't make sense to start something from scratch.

Does that make it clearer?

tonynajjar commented:
Alright. I still need to define polygons the traditional way and they need to have an action_type. I don't want them to affect cmd_vel, that's why I was proposing an extra action_type. Any counter-propositions?

SteveMacenski commented:
Doesn't this not really solve your problem since its not on a timer? 

If you're introducing a timer, there's not really a need to play with the action type stuff because, as you mentioned, there is no action to be triggered. You're just publishing that there is / is not a violation of the polygons.

tonynajjar commented:
I think there is a small misunderstanding. I need to create **new** polygons (the ones that don't cause any action) but polygons need an action_type otherwise we get an error.

SteveMacenski commented:
Oh got it. Then why not set the type to `DO_NOTHING` for this case? Nothing is an action that could be used for this new node / timer polygon type

tonynajjar commented:
> But I don't think adding another polygon type for "do nothing" is a great answer

That's what I meant but you weren't a big fan of it earlier😅, or maybe  you understood something else.

tonynajjar commented:
Alright so I have a good idea on how to get started, let's discuss it further after my initial changes!

tonynajjar commented:
I changed the PR, could you check and tell me if you have conceptual concerns? (I know there are a lot of formatting and semantics issues)

SteveMacenski commented:
You have a linting problem in the CI results. I'm still wondering if this is really the right node for it. I agree this should be in this **package** but the package can have multiple executable nodes. Your use-case of this as a polygon detector rather than  mobile-base collision monitor is sufficiently different I feel this should outright be a different node as to not mix psuedo-safety systems with application systems. While I don't now believe there's any technical reason adding more features to this node would necessarily cause a problem -- it seems architecturally like a dangerous choice. 

Plus, the second node would be much simpler, since you don't need cmd_vel or any of that stuff. 

Lets pause this for a couple of days until #3504 is merged since that's going to cause conflicts here.

tonynajjar commented:
> You have a linting problem in the CI results. 

Yeah I know, will worry about this at the end

> . I'm still wondering if this is really the right node for it. I agree this should be in this package but the package can have multiple executable nodes.

I tend to agree, I'll investigate a separate node and report back

tonynajjar commented on nav2_collision_monitor/src/detector_node.cpp:
```diff
@@ -0,0 +1,324 @@
+    source->getData(curr_time, collision_points);
+  }
+
+  std::unique_ptr<std_msgs::msg::Bool> trigger_msg =
+    std::make_unique<std_msgs::msg::Bool>();
+  for (std::shared_ptr<Polygon> polygon : polygons_) {
+    trigger_msg->data = polygon->getPointsInside(collision_points) > polygon->getMaxPoints();
+  }
+
+  trigger_pub_->publish(std::move(trigger_msg));
```
TODO; custom msg to publish state of all polygons

tonynajjar commented:
I made a new node, can you check if any **architectural** objections before I invest more time?

> Some of the polygon configuration stuff might even be able to be pulled out into a utils file and shared among both.

yes agree, there is quite some duplicated code.

SteveMacenski commented:
You should use that other PR's message type. You can set the action type to DO_NOTHING. That way you get a name!

tonynajjar commented:
That still doesn't solve having multiple polygons triggered at the same time. The msg needs an array

SteveMacenski commented on nav2_collision_monitor/src/polygon.cpp:
```diff
@@ -258,6 +258,8 @@ bool Polygon::getCommonParameters(std::string & polygon_pub_topic)
       action_type_ = SLOWDOWN;
     } else if (at_str == "approach") {
       action_type_ = APPROACH;
+    } else if (at_str == "do_nothing") {
```
`none` would be cleaner looking

tonynajjar replied:
to be clear, we're talking only about the param string right? We agree to keep the type as DO_NOTHING and not create an extra type?
About the actual, I have no strong opinion, let me know.

SteveMacenski replied:
Yeah, just the parameter name. `notify` would be fine too

AlexeyMerzlyakov replied:
> We agree to keep the type as DO_NOTHING and not create an extra type?

I think, that should be enough in current design: if we are not going to support approach notification type, it could just re-use `DO_NOTHING` action type for notification in `CollisionDetector` (otherwise, this better to take `STOP` and `APPROACH` values).

SteveMacenski replied:
Just needs the docs / readme update on adding the new param option

SteveMacenski commented on nav2_collision_monitor/CMakeLists.txt:
```diff
@@ -105,6 +126,7 @@ endif()
 
 ament_export_include_directories(include)
 ament_export_libraries(${library_name})
+ament_export_libraries(${library_name_detector})
```
You can include this on the same export call above, just add this to the list

SteveMacenski commented on nav2_collision_monitor/CMakeLists.txt:
```diff
@@ -38,7 +38,9 @@ set(dependencies
 )
 
 set(executable_name collision_monitor)
+set(executable_name_detector detector)
```
lets do collision_detector, add some more context to the executable name

SteveMacenski commented on nav2_collision_monitor/src/detector_node.cpp:
```diff
@@ -0,0 +1,324 @@
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+  // Creating timer
+  timer_ = this->create_wall_timer(
```
Move this to on activate + `timer_.reset()` on deactivate. Then you can remove the `process_active_` variable since `process()` could literally not be called if not active.

SteveMacenski reviewed (changes requested):
I read this in the spirit meant for design -- No issues. Looks great to me beyond the obvious points for removing duplication with utils. It will need test coverage though

AlexeyMerzlyakov commented:
I see new architectural update: `collision_detector` acting in parallel with `collision_monitor` node.
There are some repeating parts in the code, but this approach seems much more straightforward and reasonable, instead of trying to fit detection part into existing collision_monitor logic.

> But the polygon will still need an action_type and I think it's confusing to have a polygon with an action_type, e.g. "approach" and have the do_nothing option as well.

Yes, this is still confusing, for me too. The polygon of "STOP" type that is not stopping might seem strange. This is the main flaw of the approach with polygon action disabling parameter, for the option when (re)using the same `collision_monitor` for detecting obstacles.

> I'm not sure I understand this. What do you mean by "utilize the full logic of STOP and APPROACH situations" and why would this logic, whatever it is, be useful for a polygon that should do no action.
Also, about this: "while DO_NOTHING new polygon type will just repeat only STOP logic.". Why would a DO_NOTHING polygon use the STOP logic? It would have it's own logic

I meant the case, when the robot is 2 seconds away from (before) the collision. This might be useful case to warn external operator or S/W node about possible collision in near future. This is similar to APPROACH model, but without approach itself. Something like DO_NOTHING - is a STOP-like points detection without stop.

-----------------------

To summarize up, the way with `collision_detector` node, detecting the collisions (and in the future approaches to collisions) should be a good architectural choice to stop at.

AlexeyMerzlyakov commented on nav2_collision_monitor/src/detector_node.cpp:
```diff
@@ -0,0 +1,324 @@
+    source->getData(curr_time, collision_points);
+  }
+
+  std::unique_ptr<std_msgs::msg::Bool> trigger_msg =
+    std::make_unique<std_msgs::msg::Bool>();
```
Implicitly initialize `trigger_msg->data` with `false`, just for the case

AlexeyMerzlyakov reviewed (commented):
Overall, it is a good decision: to make a separate `collision_detector` node to not mix the action code and do not make it hairy.
This however, this does not cover the case, when the robot is away less on `N` seconds from the collision (`APPROACH` model analogue). `process()` should be called from `cmd_vel_in` callback for that case to know current robot speed for collision prediction.
This is probably, the talk of separate ticket, when(if) we will a need in such functionality, I think.

Otherwise, it seems a reasonable approach for me.

tonynajjar commented on nav2_collision_monitor/src/detector_node.cpp:
```diff
@@ -0,0 +1,324 @@
+  std::unique_ptr<std_msgs::msg::Bool> trigger_msg =
+    std::make_unique<std_msgs::msg::Bool>();
+  for (std::shared_ptr<Polygon> polygon : polygons_) {
+    trigger_msg->data = polygon->getPointsInside(collision_points) > polygon->getMaxPoints();
```
For this node, the polygons should not have any action_type (it means nothing as no actions are implemented), so not sure we should check the type

SteveMacenski replied:
I'd just check it and log the errors if they say anything else just to let a user know there's likely a mistake

tonynajjar replied:
Done in https://github.com/ros-planning/navigation2/pull/3500/commits/ef5f8cf4222dacd20e59be0caf12872d0b596a91

tonynajjar commented on nav2_collision_monitor/src/collision_detector_node.cpp:
```diff
@@ -0,0 +1,317 @@
+  return true;
+}
+
+bool CollisionDetector::configurePolygons(
```
I propose in this function to set the action_type of the polygons to DO_NOTHING and not include the action_type yaml parameter in the docs examples. I think it would be confusing to offer an action_type yaml option when there is really only one choice.

The downside is that this makes it harder to put configurePolygons in a common utils.cpp file.

What do you think?

SteveMacenski replied:
I don't think there's a problem to set it here to that value. But I would recommend logging if its NOT either empty or `notify` (or whatever rename it) so users know there's an issue with what they're requesting the system to accomplish.

For the sake of reusing code in general utilities, it might not make sense to remove that logic here but keep the same as much as possible to be shared, then application specific code after the shared function for validation.

SteveMacenski commented:
> when the robot is away less on N seconds from the collision (APPROACH model analogue). process() should be called from cmd_vel_in callback for that case to know current robot speed for collision prediction.

This has nothing to do with cmd_vel.

SteveMacenski commented:
@tonynajjar its been a hot minute here, what's the good word?

AlexeyMerzlyakov commented:
> This has nothing to do with cmd_vel.

I agree, if we are not going to support approaching model for `CollisionDetector`.

AlexeyMerzlyakov commented on nav2_collision_monitor/src/collision_detector_node.cpp:
```diff
@@ -0,0 +1,317 @@
+  tf_listener_ = std::make_shared<tf2_ros::TransformListener>(*tf_buffer_);
+
+  state_pub_ = this->create_publisher<nav2_msgs::msg::CollisionDetectorState>(
+    "~/state", rclcpp::SystemDefaultsQoS());
```
"~/state" is better to be parameterized, similarly to how it made in `collision_monitor_node.cpp`

SteveMacenski replied:
^ I don't necessarily think it needs to be parameterized, but it needs to be made just `state` without the `~/`

AlexeyMerzlyakov replied:
Okay, in CollisionDetector logic, the state is mandatory parameter, opposite of CollisionMonitor logic (where this parameter is being optional), so I would agree that it could be hard-coded.
However, to be aligned with https://github.com/ros-planning/navigation2/commit/b8d077eebb7755ae3b7510ae7a993c499c8bed6c, we might need to rename this parameter to something like `collision_detector_state`, or rename back `collision_monitor_state` -> to just `state` there (second option seems better for me).

tonynajjar replied:
> but it needs to be made just state without the ~/

Why though? I personally would have named both topics "~/state" for both CD and CM, so the result is "/collision_detector/state" and "/collision_monitor/state". I usually like to have the node name in the topic to know instinctively where it comes from, especially for such a generic topic like "state".
Also I usually don't parametrize topic names as they can easily be remapped if the user wants to.

Anyway no strong opinion, I'm fine with calling it "collision_detector_state" as well. Let me know

> second option seems much easier and better for me

That would break things for someone already using the CM with the default configuration, I don't think it's necessary given the other options.

AlexeyMerzlyakov reviewed (commented):
I've returned back with the next round of review. Now it seems we are on the same page with the design and main change. There are some small items are remained to update, but overall the changes are look to be quite formed and finished.

SteveMacenski commented:
Any updates here? Its been a  hot minute since things progressed and I'd really like to have this merged in!

tonynajjar commented:
Hey, this fell out of priority for us but I'll try to find some free time to continue working on it

SteveMacenski commented:
@AlexeyMerzlyakov please re-review

SteveMacenski commented on nav2_collision_monitor/src/collision_detector_node.cpp:
```diff
@@ -0,0 +1,317 @@
+
+  // Creating timer
+  timer_ = this->create_wall_timer(
+      std::chrono::duration<double>{1.0 / frequency_},
```
Prob a linting issue once it builds

SteveMacenski commented on nav2_collision_monitor/src/collision_detector_node.cpp:
```diff
@@ -0,0 +1,317 @@
+CollisionDetector::on_shutdown(const rclcpp_lifecycle::State & /*state*/)
+{
+  RCLCPP_INFO(get_logger(), "Shutting down");
+
```
Remove space

SteveMacenski replied:
@tonynajjar still a thing :-)

SteveMacenski reviewed (commented):
I broadly approve after Alexey looks at the logic beyond what I've mentioned here as nitpicks and still pending above.

The next steps in my mind are updates to the readme for the new node + `none` mode option and the complimentary updates to the website. This also needs some test coverage. But, its a great contribution and I want to see this in!

tonynajjar commented:
It's not ready or tested yet, will ping when it is :D

AlexeyMerzlyakov reviewed (commented):
Overall, the main code change seems OK for me.
There are the following review items from above are remain to be fixed:
- https://github.com/ros-planning/navigation2/pull/3500#discussion_r1150289114
  Please use `create_timer()` instead
- https://github.com/ros-planning/navigation2/pull/3500#discussion_r1150332021
  Add a warning if some other action_type (other than `none`) is used
- https://github.com/ros-planning/navigation2/pull/3500#discussion_r1196213505
  Parameter renaming discussion
- https://github.com/ros-planning/navigation2/pull/3500#discussion_r1245563070 and   https://github.com/ros-planning/navigation2/pull/3500#discussion_r1245563285 code style changes

Also please:
- Don't forget to add your copyrights to the `collision_detector_node.hpp` and `collision_detector_node.cpp`.
  Moreover, I think, `collision_detector_main.cpp` main file should contain only your copyrights, since there are no any specific code/algorithms logic derived from any other files.
- The latest `collision_monitor_node.launch.py` launch-file could be copied to use CollisionDetector as well, or changed to have an option to run CM or either CD.
- Please add the testcases covering new Collision Detector functionality.
- Documentation updates are required:
  * New Collision Detector node part into Iron->Jazzy Migration guide
  * Collision Monitor Configuration guide should be updated with 1. common information about Collision Detector and 2. CD parameters descriptions. 
  * Also, I believe, this common information about CD should be added to `nav_collision_monitor/README.md`

tonynajjar commented:
I think I addressed all the comments but I'm having trouble setting up a rolling environment to test this.... gazebo keeps crashing. Will update once I manage to test

SteveMacenski commented:
OK! Last things:
- Test coverage of the new class functionality 
- In the README bring up this new feature and how its useful / meant for. Maybe another picture if you think it would aid in clarity with the other pictures
- Usual documentation updates on navigation.ros.org for configuration, migration

I glanced over the `configurePolygons` and `configureSources` methods, but I'll leave that to Alexey to nitpick as the original author of that structure. I'll say the spaces like:

```

      if (source_type == "scan") {
        std::shared_ptr<Scan> s = std::make_shared<Scan>(
          node, source_name, tf_buffer_, base_frame_id, odom_frame_id,
          transform_tolerance, source_timeout, base_shift_correction);

        s->configure();

        sources_.push_back(s);
```

Seem odd to me, but if that's consistent with the other files, then you can leave it. You can move this to be a full non-draft PR now I think!

AlexeyMerzlyakov commented on nav2_collision_monitor/launch/collision_detector_node.launch.py:
```diff
@@ -0,0 +1,150 @@
+#!/usr/bin/env python3
+
+# Copyright (c) 2022 Samsung R&D Institute Russia
```
Change copyrights to yours

AlexeyMerzlyakov commented on nav2_collision_monitor/include/nav2_collision_monitor/collision_detector_node.hpp:
```diff
@@ -0,0 +1,159 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef NAV2_COLLISION_MONITOR__DETECTOR_NODE_HPP_
+#define NAV2_COLLISION_MONITOR__DETECTOR_NODE_HPP_
```
`NAV2_COLLISION_MONITOR__DETECTOR_NODE_HPP_` -> `NAV2_COLLISION_MONITOR__COLLISION_DETECTOR_NODE_HPP_` according to naming rules

AlexeyMerzlyakov commented on nav2_collision_monitor/include/nav2_collision_monitor/collision_detector_node.hpp:
```diff
@@ -0,0 +1,159 @@
+
+}  // namespace nav2_collision_monitor
+
+#endif  // NAV2_COLLISION_MONITOR__DETECTOR_NODE_HPP_
```
Ditto

AlexeyMerzlyakov commented on nav2_collision_monitor/include/nav2_collision_monitor/collision_detector_node.hpp:
```diff
@@ -0,0 +1,159 @@
+  std::vector<std::shared_ptr<Source>> sources_;
+
+  /// @brief collision monitor state publisher
+  rclcpp_lifecycle::LifecyclePublisher<nav2_msgs::msg::CollisionDetectorState>::SharedPtr state_pub_;
```
This will cause linting issues (>100 symbols in the line)

AlexeyMerzlyakov replied:
Good, but there is no need to put "\" symbol in the end of line, since it is not a macro or string line transfer

AlexeyMerzlyakov commented on nav2_collision_monitor/src/collision_detector_node.cpp:
```diff
@@ -0,0 +1,338 @@
+{
+  RCLCPP_INFO(get_logger(), "Deactivating");
+
+  // Deactivating lifecycle publishers
+  state_pub_->on_deactivate();
```
`state_pub_` is better to deactivate after the `timer_` to be reset. Otherwise we might have (potential) multi-thread situation when timer callback will be called one more time after state publisher will be activated. This indeed won't appear in our case, as Nav2 is using single-threaded ROS-executor. But anyway and just for the case, it is more clear that everything is being stopped sequentially in the right order.

AlexeyMerzlyakov commented on nav2_collision_monitor/src/collision_detector_node.cpp:
```diff
@@ -0,0 +1,338 @@
+    return false;
+  }
+
+  if (!configureSources(
```
This will probably cause uncrustify issues: `!configureSources(` -> to a new line

SteveMacenski replied:
This is actually proper and fine. I actually prefer this for my personal readability, but they're both equally valid I believe.

AlexeyMerzlyakov replied:
Yes, I've checked with uncrustify and it indeed does not cause and issue. Closing item

AlexeyMerzlyakov reviewed (commented):
The main change is almost done (only my nitpick comments are below). Testcases + documentation & Readme.md update are required.

tonynajjar commented:
Still tests and some documentation is missing, not ready for full review

SteveMacenski commented:
Let me know when you want me to take a look!

tonynajjar commented:
Feel free to take a look already, last thing I did was adding tests and fix some PR comments. I think mostly documentation is left but I'll continue this early next week

AlexeyMerzlyakov commented on nav2_collision_monitor/README.md:
```diff
@@ -66,3 +67,9 @@ The following notes could be made:
+## Collision Detector
+
+Another node exists in the nav2_collision_monitor package called the Collision Detector. This node works similarly to the collision monitor
+except that it does not affect the robot's velocity. It will only inform that data from the configured sources has been detected within the configured polygons via message to the `/collision_detector_state` topic. 
```
I think, this is also important to tell that unlike Collision Monitor that uses different behavior models, Collision Detector does not apply any of them. It detects the points inside given polygon instead, and reports back to user.

AlexeyMerzlyakov commented on nav2_collision_monitor/README.md:
```diff
@@ -66,3 +67,9 @@ The following notes could be made:
  * More points mean lower performance. Pointclouds could be culled or filtered before the Collision Monitor to improve performance.
+
+
+## Collision Detector
```
Also, in the beginning of the `Collision Detector` section, it could be great to add a few words about background/prerequisites for creating CD: why it is was introduced (e.g. CM only changes robot's behavior, but we might need to have an information utilized either by external module or user) and what examples of use-cases CD to cover (e.g. LED on/off, other safety devices connection, obstacle detection alarm, connection with other robots/modules, etc...)

tonynajjar replied:
I added a little more verbosity but since it's similar to collision monitor in many aspects (as stated), I'm not sure what more to add. Please give specific examples if you think it's not enough.

AlexeyMerzlyakov commented on nav2_collision_monitor/test/collision_detector_node_test.cpp:
```diff
@@ -0,0 +1,598 @@
+  RANGE = 3
+};
+
+class CollisionDetectorWrapper : public nav2_collision_monitor::CollisionDetector
```
Code coverage of the testcase is `92.8%` which is less than average `~95%` on CM:
![total_code_coverage](https://github.com/ros-planning/navigation2/assets/60094858/39280523-d51b-4b7d-82e6-78e6d88571e2)

There are some cases in `CD::configurePolygons()` and `CD::configureSources()` yet to be covered:
![configurePolygons_coverage](https://github.com/ros-planning/navigation2/assets/60094858/61bfd82d-248d-4c80-a613-db2c62174aa2)
![configureSources_coverage](https://github.com/ros-planning/navigation2/assets/60094858/ebcb9c62-ae0c-49df-9af5-bf9a4c19b86e)

These cases logic could be copied from corner cases at `collision_monitor_node_test.cpp` as well (e.g. by making circle safety shape with PCL2 and range sensors published; also action type rather than "none" also could be easily checked).

AlexeyMerzlyakov reviewed (commented):
The next round of review is mostly about the testing and documentation updates, so please refer of the review items below

tonynajjar commented on nav2_collision_monitor/test/CMakeLists.txt:
```diff
@@ -22,14 +22,16 @@ ament_target_dependencies(polygons_test
+
+# Collision Detector node test
+ament_add_gtest(collision_detector_node_test collision_detector_node_test.cpp)
+ament_target_dependencies(collision_detector_node_test
   ${dependencies}
 )
-target_link_libraries(collision_monitor_node_test
-  ${library_name}
-)
+target_link_libraries(collision_detector_node_test
+  ${detector_library_name}
+)
```
oops, thanks for noticing

tonynajjar commented:
@AlexeyMerzlyakov thanks for the thorough review, can you please have a look again if all your comments have been addressed?

tonynajjar commented:
tests are failing, I'm checking why

SteveMacenski commented:
One of the unit tests is still failing :-)

SteveMacenski commented on nav2_collision_monitor/include/nav2_collision_monitor/collision_detector_node.hpp:
```diff
@@ -0,0 +1,159 @@
+  std::vector<std::shared_ptr<Source>> sources_;
+
+  /// @brief collision monitor state publisher
+  rclcpp_lifecycle::LifecyclePublisher<nav2_msgs::msg::CollisionDetectorState>::SharedPtr \
```
Don't use C-style line extenders. I personally like them, but its an anti-pattern in modern C++ :cry:

SteveMacenski replied:
Otherwise LGTM. I didn't pay too much attention to the specific logic but I trust that Alexey is looking that over with you and I'll just verify it once you're both happy with what's there. This seems very close to shipping though to my eye

SteveMacenski commented on nav2_collision_monitor/README.md:
```diff
@@ -66,3 +67,10 @@ The following notes could be made:
+
+Another node exists in the nav2_collision_monitor package called the Collision Detector. This node works similarly to the collision monitor except that it does not affect the robot's velocity. It will only inform that data from the configured sources has been detected within the configured polygons via message to the `collision_detector_state` topic. It is important to note that unlike the Collis ...
+
+The Collision Detector was introduced for the use cases where the robot's velocity should not be affected by the collision monitor, but the user still wants to be informed about the detected obstacles and act upon it in different ways; e.g. blink LEDs or sound an alarm when the robot is close to an obstacle.
```
A little more verbosity here might be nice

AlexeyMerzlyakov reviewed (commented):
Minor updates. Some previous comments are still not resolved. I'll be OOO next week, so on WW32 I'll be back and make the finish-line review. Before this, @tonynajjar, please open a new PR in navigation.ros.org repository with Migration Guide / parameters configuration / CM update.

tonynajjar commented:
https://github.com/ros-planning/navigation.ros.org/pull/451

tonynajjar commented:
https://app.circleci.com/pipelines/github/ros-planning/navigation2/9813/workflows/3fdd3100-a4b0-46c0-ab3f-881de07fb1f3/jobs/31374/parallel-runs/0/steps/0-117

How do I solve this?

SteveMacenski commented:
> [0.000s] [error]: Cannot find '/opt/overlay_ws/src/navigation2/nav2_collision_monitor/src/main.cpp'

I think you just renamed the files so the cache isn't quite right. If you were to increase the `v14` (or maybe `v15`) in the circlci file (there should be 3 instances) then that'll create a new cache for the change in file locales that its looking for.

tonynajjar commented:
> > [0.000s] [error]: Cannot find '/opt/overlay_ws/src/navigation2/nav2_collision_monitor/src/main.cpp'
> 
> I think you just renamed the files so the cache isn't quite right. If you were to increase the `v14` (or maybe `v15`) in the circlci file (there should be 3 instances) then that'll create a new cache for the change in file locales that its looking for.

bumped from v15 to v16, let's see

SteveMacenski commented:
Works now, just retriggering due to a (increasingly annoying) flaky test - since it would be good to see the actual code coverage report pop up. Though you can find it here: https://app.codecov.io/gh/ros-planning/navigation2/tree/collision-monitor-publish-action-main/nav2_collision_monitor%2Fsrc

It looks like you could cover a couple of extra things that would button it up. Normally for just missing a few lines I wouldn't care (since the last few lines are usually the hard ones) but the ones you have missing are pretty straight forward to test:
- Use a "circle" type for a dummy test, not just polygon
- Incorrectly set to a type other than DO_NOTHING to test your error condition that it should fail -- actually that hsouldn't just log a warning that should fail the configuration since something is critically wrong!
- Test pointcloud source, not just range and scan

These just need to be minimal tests just to exercise things, since it shares so much with the collision monitor. 
-

tonynajjar commented:
> Use a "circle" type for a dummy test, not just polygon

hmm, it's weird that the [coverage](https://app.codecov.io/gh/ros-planning/navigation2/blob/collision-monitor-publish-action-main/nav2_collision_monitor%2Fsrc%2Fcollision_detector_node.cpp#L201) shows that a Circle creation is never tested, it should be in the test called `testCircleDetection`

> actually that hsouldn't just log a warning that should fail the configuration since something is critically wrong!

okay fixed in latest commit

> Test pointcloud source, not just range and scan

I tried that for 2 hours, the test just kept failing in CI but worked locally so I gave up. Giving it another shot in the latest commit

SteveMacenski reviewed (approved):
I'm happy if Alexey is

AlexeyMerzlyakov reviewed (approved):
All review items were resolved by now. Local `nav2_collision_monitor` testing passes. `collision_detector_node.cpp` coverage = `99.4%` with total code coverage for CM package = `97.4%` which is great.
I am OK to merge this change to master (docs are in review now, but to be on finish line).

SteveMacenski commented:
Thanks for sticking with it :-)
